pandas/core/arrays/datetimes.py

```diff
@@ -239,56 +239,39 @@ def _generate_range(cls, start, end, periods, freq, tz=None,
         start, end, _normalized = _maybe_normalize_endpoints(start, end,
                                                              normalize)
 
-        tz, inferred_tz = _infer_tz_from_endpoints(start, end, tz)
+        tz, _ = _infer_tz_from_endpoints(start, end, tz)
 
-        if hasattr(freq, 'delta') and freq != Day():
-            # sub-Day Tick
-            if inferred_tz is None and tz is not None:
-                # naive dates
-                if start is not None and start.tz is None:
-                    start = start.tz_localize(tz, ambiguous=False)
+        # Make sure start and end are timezone localized if:
+        # 1) freq = a Timedelta-like frequency (Tick)
+        # 2) freq = None i.e. generating a linspaced range
+        if isinstance(freq, Tick) or freq is None:
+            localize_args = {'tz': tz, 'ambiguous': False}
+        else:
+            localize_args = {'tz': None}
 
-                if end is not None and end.tz is None:
-                    end = end.tz_localize(tz, ambiguous=False)
+        if tz is not None:
+            # Localize the start and end arguments
+            if start is not None and start.tz is None:
```

Comment:
you could use ``_maybe_localize_point(start, getattr(start, 'tz' None))`` ?

Reply:
Hmm I think that should work!